snapshotter/devmapper/devmapper.go

```diff
@@ -0,0 +1,320 @@
+		return complete(ctx, trans, err)
+	}
+
+	return complete(ctx, trans, nil)
+}
+
+func (dm *Snapshotter) Remove(ctx context.Context, key string) error {
+	log.G(ctx).WithField("key", key).Debug("remove")
+
+	ctx, trans, err := dm.store.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	snapID, _, err := storage.Remove(ctx, key)
+	if err != nil {
+		return complete(ctx, trans, err)
+	}
+
+	deviceName := dm.getDeviceName(snapID)
+	if err := dm.pool.RemoveDevice(deviceName); err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to remove device")
+		return complete(ctx, trans, err)
+	}
+
+	return complete(ctx, trans, nil)
+}
+
+func (dm *Snapshotter) Walk(ctx context.Context, fn func(context.Context, snapshots.Info) error) error {
+	log.G(ctx).Debug("walk")
+
+	ctx, trans, err := dm.store.TransactionContext(ctx, false)
+	if err != nil {
+		return err
+	}
+
+	defer trans.Rollback()
+	return storage.WalkInfo(ctx, fn)
+}
+
+func (dm *Snapshotter) Close() error {
+	log.L.Debug("close")
+
+	var result *multierror.Error
+
+	if err := dm.store.Close(); err != nil {
+		result = multierror.Append(result, err)
+	}
+
+	if err := dm.pool.Close(context.Background(), false, false); err != nil {
+		result = multierror.Append(result, err)
+	}
+
+	return result.ErrorOrNil()
+}
+
+func (dm *Snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	ctx, trans, err := dm.store.TransactionContext(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+
+	snap, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+	if err != nil {
+		return nil, complete(ctx, trans, err)
+	}
+
+	if len(snap.ParentIDs) == 0 {
+		deviceName := dm.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating new thin device '%s'", deviceName)
+
+		deviceID, err := dm.pool.CreateThinDevice(deviceName, dm.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create thin device for snapshot %s", snap.ID)
+			return nil, complete(ctx, trans, err)
+		}
+
+		log.G(ctx).Debugf("created thin device with id %d", deviceID)
+		if err := dm.mkfs(ctx, deviceName); err != nil {
+			return nil, complete(ctx, trans, err)
+		}
+	} else {
+		parentDeviceName := dm.getDeviceName(snap.ParentIDs[0])
+		snapDeviceName := dm.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating snapshot device '%s' from '%s'", snapDeviceName, parentDeviceName)
+
+		snapDeviceID, err := dm.pool.CreateSnapshotDevice(parentDeviceName, snapDeviceName, dm.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create snapshot device from parent %s", parentDeviceName)
+			return nil, complete(ctx, trans, err)
+		}
+
+		log.G(ctx).Debugf("created snapshot device with id %d", snapDeviceID)
+	}
+
+	mounts := dm.buildMounts(snap)
+
+	// This is needed only to pass containerd's snapshotter suite tests
```

Comment:
I don't think that this is unreasonable and I wouldn't necessarily characterize it as "only needed to pass containerd's tests".  Maybe rephrase the comment?

// Remove default directories not expected by the container image